### ISSUE TITLE
[Bug]: generatePath doesn't work with optional params (v6.5.0-pre.1)

### ISSUE THREAD

**Issue by jesperjohansson** [remix-run/react-router#9736]
### What version of React Router are you using?

6.5.0-pre.1

### Steps to Reproduce

Reproduction demo: https://stackblitz.com/edit/github-kmv8on-hr7pvh?file=src/App.tsx

**Attempt 1, path becomes `/food/snacks/fruit/apple?/apples` (with the `?`)** and wrong page is shown
```jsx
<Link
  to={generatePath('food/:foodName/fruit/:fruitName?/apples', {
    foodName: 'snacks',
    fruitName: 'apple',
  })}
>
  Apples (generatePath)
</Link>
```

**Attempt 2, throws error `Missing ":fruitName" param`**
```jsx
<Link
  to={generatePath('food/:foodName/fruit/:fruitName?/apples', {
    foodName: 'snacks',
    'fruitName?': 'apple',
  })}
>
  Apples (generatePath)
</Link>
```

### Expected Behavior

Should had returned `/food/snacks/fruit/apple/apples`

### Actual Behavior

Returns `/food/snacks/fruit/apple?/apples`

**Comment by jesperjohansson** [comment]
@brophdawg11 Quick test seems to work https://stackblitz.com/edit/github-kmv8on-xzfgau?file=src/App.tsx

**Comment by brophdawg11** [comment]
Released in `6.7.0`
